Ensure that list-items that are `dict(s)` are [also] converted

jzaleski opened the pull request:
I have been running this as a monkey-patch for a while, but wanted to push this upstream.

This change ensures that `list`-items that are `dict(s)` are [also] converted during the init-process and bumps the version to 3.1.0 as this _could_ be a breaking change.

hamidnazari commented:
Hi @jzaleski 

Thanks for the patch. I understand what you're trying to do here. But why not wrap the `dict` assignment in an `OpenStruct` constructor?

```
o = OpenStruct()
o.x = {'a': 1, 'b': 2, 'c': 3}
o.y = OpenStruct({'a': 1, 'b': 2, 'c': 3})
type(o.x) # dict
type(o.y) # ostruct.openstruct.OpenStruct
```

As you mentioned this patch is a behaviour change. I also think dictionaries as types should be able to be assigned and retrieved as they are, unless explicitly converted.

Alternatively, we can give the ostruct object a property flag to automatically convert dictionaries if set, e.g. 

```
o = OpenStruct(dict_conevert=True)
o.x = {'a': 1, 'b': 2, 'c': 3}
o.y = OpenStruct({'a': 1, 'b': 2, 'c': 3})
type(o.x) # ostruct.openstruct.OpenStruct
type(o.y) # ostruct.openstruct.OpenStruct
```

Thoughts?

jzaleski commented:
Valid point w/ the auto-converting, if we make it opt-in we don't run the risk of breaking others and it will also guarantee backwards compatibility.

I am going to update the PR to add a `dict_convert` kwarg and default it to the old behavior.. I will also remove the update to the version-file so that you can do it on your end when you're ready to push a new version to _PyPi_.

jzaleski commented:
Made it so that the behavior is now opt-in. I also cleaned-up the test file for consistency and fixed a deprecation warning.

As mentioned previously, I took the opportunity to remove the version-update from my PR. I'll let you decide on that, at this point I think `3.0.2` is probably the next logical version.

Thanks for the feedback!

hamidnazari commented:
Looks great. Thanks for the changes. Please refer to the feedback above, just a tiny change and it's good to be merged. I'll do the version bump afterwards.

jzaleski commented:
Should be all set! Thanks for your help in reviewing this @hamidnazari

jzaleski commented:
Should be all set! Thanks for your help in reviewing this @hamidnazari
